xarray/core/indexes.py

```diff
@@ -1,12 +1,202 @@
+        return self.array
+
+    @property
+    def dtype(self) -> np.dtype:
+        return self._dtype
+
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        if dtype is None:
+            dtype = self.dtype
+        array = self.array
+        if isinstance(array, pd.PeriodIndex):
+            with suppress(AttributeError):
+                # this might not be public API
+                array = array.astype("object")
+        return np.asarray(array.values, dtype=dtype)
+
+    @property
+    def shape(self) -> Tuple[int]:
+        return (len(self.array),)
+
+    def equals(self, other):
+        if isinstance(other, pd.Index):
+            other = PandasIndex(other)
+        return isinstance(other, PandasIndex) and self.array.equals(other.array)
+
+    def union(self, other):
+        if isinstance(other, pd.Index):
+            other = PandasIndex(other)
+        return PandasIndex(self.array.union(other.array))
+
+    def intersection(self, other):
+        if isinstance(other, pd.Index):
+            other = PandasIndex(other)
+        return PandasIndex(self.array.intersection(other.array))
+
+    def __getitem__(
+        self, indexer
+    ) -> Union[
+        "PandasIndex",
+        NumpyIndexingAdapter,
+        np.ndarray,
+        np.datetime64,
+        np.timedelta64,
+    ]:
+        key = indexer.tuple
+        if isinstance(key, tuple) and len(key) == 1:
+            # unpack key so it can index a pandas.Index object (pandas.Index
+            # objects don't like tuples)
+            (key,) = key
+
+        if getattr(key, "ndim", 0) > 1:  # Return np-array if multidimensional
+            return NumpyIndexingAdapter(self.array.values)[indexer]
+
+        result = self.array[key]
+
+        if isinstance(result, pd.Index):
+            result = PandasIndex(result, dtype=self.dtype)
+        else:
+            # result is a scalar
+            if result is pd.NaT:
+                # work around the impossibility of casting NaT with asarray
+                # note: it probably would be better in general to return
+                # pd.Timestamp rather np.than datetime64 but this is easier
+                # (for now)
+                result = np.datetime64("NaT", "ns")
+            elif isinstance(result, timedelta):
+                result = np.timedelta64(getattr(result, "value", result), "ns")
+            elif isinstance(result, pd.Timestamp):
+                # Work around for GH: pydata/xarray#1932 and numpy/numpy#10668
+                # numpy fails to convert pd.Timestamp to np.datetime64[ns]
+                result = np.asarray(result.to_datetime64())
+            elif self.dtype != object:
+                result = np.asarray(result, dtype=self.dtype)
+
+            # as for numpy.ndarray indexing, we always want the result to be
+            # a NumPy array.
+            result = utils.to_0d_array(result)
+
+        return result
+
+    def transpose(self, order) -> pd.Index:
```

Comment:
Good catch. That's something I overlooked, probably among other things. That would make indeed sense to return a `PandasIndex`, although it might not be necessary (at least for now) as the returned `pd.Index` is later converted into a `PandasIndex` elsewhere (e.g., like in `Variable.transpose`). That might even not be desirable as currently (if I'm correct) creating a new `PandasIndex` from a `PandasIndex` re-builds the whole underlying `pd.Index`. Or alternatively we need a fastpath creation for this specific case.